sys/net/skald/skald.c

```diff
@@ -0,0 +1,141 @@
+#include <stdint.h>
+
+#include "assert.h"
+#include "random.h"
+#include "luid.h"
+
+#include "net/netdev/ble.h"
+#include "net/skald.h"
+
+#define ENABLE_DEBUG            (0)
+#include "debug.h"
+
+#define JITTER_MIN              (0U)            /* 0ms */
+#define JITTER_MAX              (10000U)        /* 10ms */
+
+#define ADV_CHAN_NUMOF          sizeof(_adv_chan)
+#define ADV_AA                  (0x8e89bed6)    /* access address */
+#define ADV_CRC                 (0x00555555)    /* CRC initializer */
+
+static const uint8_t _adv_chan[] = SKALD_ADV_CHAN;
+
+static netdev_ble_ctx_t _ble_ctx = {
+    .aa.u32 = ADV_AA,
+    .crc = ADV_CRC,
+};
+
+static netdev_t *_radio;
+
+static void _stop_radio(void)
+{
+    netdev_ble_stop(_radio);
+    _radio->context = NULL;
+}
+
+static void _sched_next(skald_ctx_t *ctx)
+{
+    ctx->last += SKALD_INTERVAL;
+    /* schedule next advertising event, adding a random jitter between
+     * 0ms and 10ms (see spec v5.0-vol6-b-4.4.2.2.1) */
+    ctx->last += random_uint32_range(JITTER_MIN, JITTER_MAX);
+    /* compensate the time passed since the timer triggered last by using the
+     * current value of the timer */
+    xtimer_set(&ctx->timer, (ctx->last - xtimer_now_usec()));
+}
+
+static void _on_adv_evt(void *arg)
+{
+    skald_ctx_t *ctx = (skald_ctx_t *)arg;
+
+    /* advertise on the next adv channel - or skip this event if the radio is
+     * busy */
+    if (_radio->context != NULL) {
+
+    }
+    if ((ctx->cur_chan < ADV_CHAN_NUMOF) && (_radio->context == NULL)) {
+        _radio->context = ctx;
+        _ble_ctx.chan = _adv_chan[ctx->cur_chan];
+        netdev_ble_set_ctx(_radio, &_ble_ctx);
+        netdev_ble_send(_radio, &ctx->pkt);
+        ++ctx->cur_chan;
+    }
+    else {
+        ctx->cur_chan = 0;
+        _sched_next(ctx);
+    }
+}
+
+static void _on_radio_evt(netdev_t *netdev, netdev_event_t event)
+{
+    (void)netdev;
+
+    if (event == NETDEV_EVENT_TX_COMPLETE) {
+        skald_ctx_t *ctx = _radio->context;
+        _stop_radio();
+        xtimer_set(&ctx->timer, 150);
+    }
+}
+
+void skald_init(netdev_t *dev)
+{
+    assert(dev);
+
+    _radio = dev;
+    _radio->event_callback = _on_radio_evt;
+    _radio->driver->init(_radio);
+}
+
+void skald_adv_start(skald_ctx_t *ctx)
+{
+    assert(ctx);
+
+    /* make sure the given context is not advertising at the moment */
+    skald_adv_stop(ctx);
+
+    /* initialize advertising context */
+    ctx->timer.callback = _on_adv_evt;
+    ctx->timer.arg = ctx;
+    ctx->last = xtimer_now_usec();
+    ctx->cur_chan = 0;
+    ctx->pkt.flags = (BLE_ADV_NONCON_IND | GORM_LL_FLAG_TXADD);
+
+    /* start advertising */
+    xtimer_set(&ctx->timer, SKALD_INTERVAL);
+}
+
+void skald_adv_stop(skald_ctx_t *ctx)
+{
+    assert(ctx);
+
+    xtimer_remove(&ctx->timer);
+    if (_radio->context == (void *)ctx) {
+        _stop_radio();
+    }
+}
+
+void skald_generate_random_addr(uint8_t *buf)
+{
+    assert(buf);
+
+    luid_get(buf, BLE_ADDR_LEN);
```

Comment:
AFAIK, BLE addresses are MAC addresses as defined in IEEE 802. Please at least set the universal/local bit to local and the individual/group bit to individual after calling this:

```C
    buf[0] &= 0x01;
    buf[0] |= 0x02;
```

Reply:
fixed.